Complete rewrite, added support for hex IDs and auth keys

martinber opened the pull request:
If you want you can merge the branches. I have added support for hexadecimal IDs and now the URL used to download the tiles is the same as the one in the KML, so now we can download tiles with auth keys. For example this gigapan: http://www.gigapan.com/gigapans/0d54f8d3323a60308eb2ff9831edfe4a has an hex ID and an auth key
I rewrote the code because I wanted to learn better python, so if you want you can leave this branch as a fork

Commit:
Complete rewrite, added support for hex IDs and auth keys

Commit:
Add license

Commit:
Fixed headers in README

FlestRus commented:
привет.

> If you want you can merge the branches. I have added support for hexadecimal IDs and now the URL used to download the tiles is the same as the one in the KML, so now we can download tiles with auth keys. For example this gigapan: http://www.gigapan.com/gigapans/0d54f8d3323a60308eb2ff9831edfe4a has an hex ID and an auth key I rewrote the code because I wanted to learn better python, so if you want you can leave this branch as a fork

Hi, friend.
thanks for the script.
but I found a couple of problems with it

first and foremost
sometimes the site responds with an error and the download is interrupted, you have to restart the script again, a lot of time is lost 


Traceback (most recent call last):
  File "C:\gigapandown\gigapanDownloader.py", line 95, in <module>
    h = urlopen(url)
  File "C:\Python27\lib\urllib.py", line 87, in urlopen
    return opener.open(url)
  File "C:\Python27\lib\urllib.py", line 213, in open
    return getattr(self, name)(url)
  File "C:\Python27\lib\urllib.py", line 350, in open_http
    h.endheaders(data)
  File "C:\Python27\lib\httplib.py", line 1038, in endheaders
    self._send_output(message_body)
  File "C:\Python27\lib\httplib.py", line 882, in _send_output
    self.send(msg)
  File "C:\Python27\lib\httplib.py", line 844, in send
    self.connect()
  File "C:\Python27\lib\httplib.py", line 821, in connect
    self.timeout, self.source_address)
  File "C:\Python27\lib\socket.py", line 557, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
IOError: [Errno socket error] [Errno 11001] getaddrinfo failed



2

are you using jpg files
but it has a limit of 50,000 pixels, which is not enough.
i fixed it to '.tif' and the problem was solved

3

while merging files, the script uses a LOT of space on the C drive
you need the ability to specify another disk for the cache.
I don't know python language. but was able to partially solve this problem
subprocess.call('"'+imagemagick+'" -define registry:temporary-path=E:\tmp -depth 8 '+

now when I run the script from the E drive via the Windows command line, the cache is created in the E drive

but the script after a successful merging of the lines still gives some kind of incomprehensible error
montage.exe: unable to open image `mp': No such file or directory @ error/blob.c/OpenBlob/2761.
montage.exe: no decode delegate for this image format `' @ error/constitute.c/ReadImage/504.
OK


if you could fix these errors and most importantly solve the problem with the disconnect, it would be very cool
I don’t even dream about loading into multiple threads  ))

martinber commented:
Hello! Its been a very long time since I worked with this and I dont remember anything (and to be fair when I modified DeniR's code I didn't know anything about python haha).

1. If the error happens sometimes, I would add some kind of loop to retry a maximum of e.g. 10 times. (tell me if you need help with this)
2. I see the output format is TIP and not JPG, I don't understand where the problem comes from
3. I think this is maybe a problem of the "montage" program and not of this code, or maybe is related to the fact you changed something from JPG to TIF? Maybe you can try something from here: https://stackoverflow.com/questions/37594157/imagemagick-no-decode-delegate-for-this-image-format-error-constitute-c-re

FlestRus commented:
> Hello! Its been a very long time since I worked with this and I dont remember anything (and to be fair when I modified DeniR's code I didn't know anything about python haha).
> 
> 1. If the error happens sometimes, I would add some kind of loop to retry a maximum of e.g. 10 times. (tell me if you need help with this)
> 2. I see the output format is TIP and not JPG, I don't understand where the problem comes from
> 3. I think this is maybe a problem of the "montage" program and not of this code, or maybe is related to the fact you changed something from JPG to TIF? Maybe you can try something from here: https://stackoverflow.com/questions/37594157/imagemagick-no-decode-delegate-for-this-image-format-error-constitute-c-re

1 yes. help would be nice)
The error happens at different intervals. sometimes every 10 minutes, sometimes every 2 hours...
eg: if download error -> wait 60sec -> retry download

2 problem with "lines"
by default, lines are created in the format  jpg
jpg have a limit of 50,000 pixels

113     str(photo_id)+'/line-'+ str(lineNo) +'.jpg',
121     str(photo_id)+'/line-*.jpg ' +

3
yes, it is a problem in the "montage" 
therefore, she needs to specify that she would create a cache on another disk if there is not enough space on disk С

-define registry:temporary-path=E:\tmp 
(found this solution on their forum)

while running with 200,000 tiles, it uses > 200 GB of space
so I had to change the cache path.
it's scary to imagine what will happen when I download a panorama consisting of 1,000,000 tiles)

martinber committed:
Python 3 rewrite

martinber committed:
Join tiles by lines as made by Maciej Krok

martinber committed:
Retry downloads, documentation and more

martinber commented:
I rewrote it for Python 3, made it to use the .tif format and implemented to do 5 retries without waiting. If you want something in particular (add waiting time between retries, add parallel download) add an issue here and I will can do it the next days/weeks https://github.com/martinber/Gigapan-Downloader

FlestRus commented:
wow!!! very cool
I need a couple of days to test it)

1 it would be nice to add a cache folder setting for "montage" and not just folders for output file and tiles :)
tiles do not weigh much, but
when stitching lines, "montage" uses a LOT of space, I have 180gb free on drive C, and I had to specify drive E because lines and their merging required more than 200 GB, this is with 200 000 tiles, for panoramas with 1000 000 tiles it will probably be required about 1TB

-define registry:temporary-path=
this is the solution from the "montage" forum

2 about multiple download threads.
I think you don't need to waste your time on this. I tried to download panorama using "SAS Planet"
it has a parallel download option, i tried 4-10-20 threads, it's useless...  max speed 1-2 tiles/sec.
I think the problem is on the server side

FlestRus commented:
fitst try  with .cmd
"C:\Users\Flest\AppData\Local\Programs\Python\Python310\python.exe" "E:\gigapandown\gigapanNEW.py" 16181 -l 3 --retries 30 --out_format psb

The system cannot find the path specified.
Traceback (most recent call last):
  File "E:\gigapandown\gigapanNEW.py", line 208, in <module>
    main(args.img_name, args.req_level, args.out_folder, args.out_format, args.dry_run, args.retries)
  File "E:\gigapandown\gigapanNEW.py", line 179, in main
    download_tiles(out_folder, output_format, img_id, tiles_url, height_tiles, width_tiles, tile_size_px, retries)
  File "E:\gigapandown\gigapanNEW.py", line 127, in download_tiles
    subprocess.run([
  File "C:\Users\Flest\AppData\Local\Programs\Python\Python310\lib\subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['C:\\Program Files\\ImageMagick-6.8.5-Q16\\montage.exe -depth 8 -geometry 256x256+0+0                 -mode concatenate -tile 8x                 downloaded/16181/0000-*.jpg                 downloaded/16181/line-0000.psb']' returned non-zero exit status 1.

as I understand it, this is some kind of feature of the Windows command line and the fact that I'm running the script from .cmd

martinber commented:
Sorry, I have the problem that I don't use the program anymore, I don't use Windows and I have only limited mobile data these months. Therefore I don't think I will test in big images or work much more on this.

Regarding the `-define registry:temporary-path=` I can add the option. But I don't understand the error you listed. It says "The system cannot find the path specified" and then it just says "returned non-zero exit status 1" so I don't know what is happening

FlestRus commented:
thank you for wasting your time for me and your mobile traffic (sorry about that)

1 error from the last post is related to spaces in the path to "montage.exe"
i changed to ( montage_command = "C:\\montage.exe" ) and now the script works

2 is funny, but there are MANY times fewer problems with loading on python 3 than on python 2. so few that manually restarting the script does not seem to be a problem, since errors are very rare.
and python 3 is faster, loading tiles 1.5 times faster than python 2 maybe 2 times

3 about "retries"
without the time.sleep() delay they don't help as even 30 retries are done in 0.01sec :) and all 30 retries gets the same loading error.  now I added time.sleep(5) but so far unfortunately I haven't seen any loading errors. so I don't know if my solution helped, maybe there just haven't been any loading errors yet

The last time I got a loading error was before adding time.sleep(5).
this is how it looked https://anotepad.com/notes/wyajw3xb

4 I don't know how to change someone else's code here, and I'm afraid of breaking something.
and here is my version of the script, I don’t know python, but I was make a few changes https://anotepad.com/notes/ssyjkm9k
i also added (-define registry:temporary-path=) now the script does not use all the free space on the C drive (windows)

I keep testing.

teliseo committed:
Fix test which was broken by previous commit

teliseo committed:
gigapan.org -> gigapan.com

teliseo committed:
Fix string continuation

teliseo committed:
Use requests library (which supports HTTP keep-alive), doubling download speed

martinber committed:
Merge pull request #1 from teliseo/requests_lib

 Use requests library (which supports HTTP keep-alive), doubling download speed
